File: admin-frontend/src/api/dict.ts
import { api } from './client'

// ---------- Types ----------

export interface DictType {
  id: number
  code: string
  name_zh: string
  name_en: string
  is_system: boolean
  created_at: string
}

export interface DictItem {
  id: number
  type_id: number
  code: string
  label_zh: string
  label_en: string
  sort: number
  enabled: boolean
  created_at: string
}

export interface UpsertDictType {
  code: string
  name_zh: string
  name_en: string
}

export interface UpsertDictItem {
  code: string
  label_zh: string
  label_en: string
  sort?: number
  enabled?: boolean
}

// ---------- Public dict (already in api/dict.ts) ----------

export interface DictItemPublic {
  code: string
  label_zh: string
  label_en: string
  sort: number
}

export async function getByType(typeCode: string) {
  const { data } = await api.get<DictItemPublic[]>(`/dict/${encodeURIComponent(typeCode)}`)
  return data
}

// ---------- Admin ----------

export async function adminListTypes() {
  const { data } = await api.get<DictType[]>('/admin/dict/types')
  return data
}

export async function adminCreateType(input: UpsertDictType) {
  const { data } = await api.post<DictType>('/admin/dict/types', input)
  return data
}

export async function adminUpdateType(id: number, input: UpsertDictType) {
  const { data } = await api.put<DictType>(`/admin/dict/types/${id}`, input)
  return data
}

export async function adminDeleteType(id: number) {
  await api.delete(`/admin/dict/types/${id}`)
}

export async function adminListItems(typeId: number) {
  const { data } = await api.get<DictItem[]>(`/admin/dict/types/${typeId}/items`)
  return data
}

export async function adminCreateItem(typeId: number, input: UpsertDictItem) {
  const { data } = await api.post<DictItem>(`/admin/dict/types/${typeId}/items`, input)
  return data
}

export async function adminUpdateItem(id: number, input: UpsertDictItem) {
  const { data } = await api.put<DictItem>(`/admin/dict/items/${id}`, input)
  return data
}

export async function adminDeleteItem(id: number) {
  await api.delete(`/admin/dict/items/${id}`)
}

File: backend/src/models/dict.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use time::OffsetDateTime;

#[derive(Debug, Clone, FromRow, Serialize)]
pub struct DictType {
    pub id: i64,
    pub code: String,
    pub name_zh: String,
    pub name_en: String,
    pub is_system: bool,
    #[serde(with = "time::serde::rfc3339")]
    pub created_at: OffsetDateTime,
}

#[derive(Debug, Clone, FromRow, Serialize)]
pub struct DictItem {
    pub id: i64,
    pub type_id: i64,
    pub code: String,
    pub label_zh: String,
    pub label_en: String,
    pub sort: i32,
    pub enabled: bool,
    #[serde(with = "time::serde::rfc3339")]
    pub created_at: OffsetDateTime,
}

/// Public-facing item — drops bookkeeping fields the frontend doesn't need.
#[derive(Debug, Clone, Serialize)]
pub struct DictItemPublic {
    pub code: String,
    pub label_zh: String,
    pub label_en: String,
    pub sort: i32,
}

impl From<&DictItem> for DictItemPublic {
    fn from(it: &DictItem) -> Self {
        Self {
            code: it.code.clone(),
            label_zh: it.label_zh.clone(),
            label_en: it.label_en.clone(),
            sort: it.sort,
        }
    }
}

#[derive(Debug, Deserialize, validator::Validate)]
pub struct UpsertDictItem {
    #[validate(length(min = 1, max = 128))]
    pub code: String,
    #[validate(length(min = 1, max = 128))]
    pub label_zh: String,
    #[validate(length(min = 1, max = 128))]
    pub label_en: String,
    pub sort: Option<i32>,
    pub enabled: Option<bool>,
}

#[derive(Debug, Deserialize, validator::Validate)]
pub struct UpsertDictType {
    #[validate(length(min = 1, max = 64))]
    pub code: String,
    #[validate(length(min = 1, max = 64))]
    pub name_zh: String,
    #[validate(length(min = 1, max = 64))]
    pub name_en: String,
}
